cpp/src/arrow/util/decimal_test.cc

```diff
@@ -728,72 +858,143 @@ TYPED_TEST_SUITE(TestDecimalToReal, RealTypes);
-  CheckDecimalToReal<double>("-9223372036854777856", 0, -9.223372036854778e+18);
+  CheckDecimalToReal<Decimal, Real>("9223372036854777856", 0, 9.223372036854778e+18);
+  CheckDecimalToReal<Decimal, Real>("-9223372036854777856", 0, -9.223372036854778e+18);
   // 2**64 - 2**11 (exactly representable in a double)
-  CheckDecimalToReal<double>("18446744073709549568", 0, 1.844674407370955e+19);
-  CheckDecimalToReal<double>("-18446744073709549568", 0, -1.844674407370955e+19);
+  CheckDecimalToReal<Decimal, Real>("18446744073709549568", 0, 1.844674407370955e+19);
+  CheckDecimalToReal<Decimal, Real>("-18446744073709549568", 0, -1.844674407370955e+19);
   // 2**64 + 2**11 (exactly representable in a double)
-  CheckDecimalToReal<double>("18446744073709555712", 0, 1.8446744073709556e+19);
-  CheckDecimalToReal<double>("-18446744073709555712", 0, -1.8446744073709556e+19);
+  CheckDecimalToReal<Decimal, Real>("18446744073709555712", 0, 1.8446744073709556e+19);
+  CheckDecimalToReal<Decimal, Real>("-18446744073709555712", 0, -1.8446744073709556e+19);
   // Almost 10**38 (minus 2**73)
-  CheckDecimalToReal<double>("99999999999999978859343891977453174784", 0,
-                             9.999999999999998e+37);
-  CheckDecimalToReal<double>("-99999999999999978859343891977453174784", 0,
-                             -9.999999999999998e+37);
-  CheckDecimalToReal<double>("99999999999999978859343891977453174784", 10,
-                             9.999999999999998e+27);
-  CheckDecimalToReal<double>("-99999999999999978859343891977453174784", 10,
-                             -9.999999999999998e+27);
-  CheckDecimalToReal<double>("99999999999999978859343891977453174784", -10,
-                             9.999999999999998e+47);
-  CheckDecimalToReal<double>("-99999999999999978859343891977453174784", -10,
-                             -9.999999999999998e+47);
+  CheckDecimalToReal<Decimal, Real>("99999999999999978859343891977453174784", 0,
+                                    9.999999999999998e+37);
+  CheckDecimalToReal<Decimal, Real>("-99999999999999978859343891977453174784", 0,
+                                    -9.999999999999998e+37);
+  CheckDecimalToReal<Decimal, Real>("99999999999999978859343891977453174784", 10,
+                                    9.999999999999998e+27);
+  CheckDecimalToReal<Decimal, Real>("-99999999999999978859343891977453174784", 10,
+                                    -9.999999999999998e+27);
+  CheckDecimalToReal<Decimal, Real>("99999999999999978859343891977453174784", -10,
+                                    9.999999999999998e+47);
+  CheckDecimalToReal<Decimal, Real>("-99999999999999978859343891977453174784", -10,
+                                    -9.999999999999998e+47);
+}
+
+// Same tests, but for Decimal256
+class TestDecimal256ToRealDouble
+    : public TestDecimalToReal<std::pair<Decimal128, double>> {};
+
+TEST_F(TestDecimal256ToRealDouble, LargeValues) {
```

Comment:
Instead of copy-pasting these test functions, why not factor them out?
(for example as `TestDecimalToReal` methods?)